src/librustc_resolve/late/diagnostics.rs

```diff
@@ -225,13 +226,31 @@ impl<'a> LateResolutionVisitor<'a, '_, '_> {
         if candidates.is_empty() && is_expected(Res::Def(DefKind::Enum, crate_def_id)) {
             let enum_candidates =
                 self.r.lookup_import_candidates(ident, ns, &self.parent_scope, is_enum_variant);
-            let mut enum_candidates = enum_candidates
-                .iter()
-                .map(|suggestion| import_candidate_to_enum_paths(&suggestion))
-                .collect::<Vec<_>>();
-            enum_candidates.sort();
 
             if !enum_candidates.is_empty() {
+                if let (PathSource::Type, Some(Expr { kind: ExprKind::Type(expr, ty), .. })) =
+                    (source, self.diagnostic_metadata.current_expr)
```

Comment:
As I mentioned in https://github.com/rust-lang/rust/pull/74211#discussion_r453065411 `current_expr` as it's calculated now can be arbitrarily far from the resolved path, so it's not suitable for this check assuming the immediate parent expression.

Reply:
I changed the approach to use the already present `current_type_ascription`. Let me know what you think.